File: src/main/java/me/partlysanestudios/partlysaneskies/dungeons/PlayerRating.java
//
// Written by Su386.
// See LICENSE for copyright and license notices.
//


package me.partlysanestudios.partlysaneskies.dungeons;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import me.partlysanestudios.partlysaneskies.PartlySaneSkies;
import me.partlysanestudios.partlysaneskies.data.pssdata.PublicDataManager;
import me.partlysanestudios.partlysaneskies.system.commands.PSSCommand;
import me.partlysanestudios.partlysaneskies.utils.ChatUtils;
import me.partlysanestudios.partlysaneskies.utils.MathUtils;
import me.partlysanestudios.partlysaneskies.utils.StringUtils;
import net.minecraftforge.client.event.ClientChatReceivedEvent;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class PlayerRating {

    private static String currentPlayer = "";

    public static HashMap<String, String> positivePatterns = new HashMap<>();

    // A map that has the <Pattern, <Cause Category, Points>>
    private static HashMap<String, HashMap<String, Integer>> playerPointCategoryMap = new HashMap<>();
    // A map that has <Player, Total Points> Map
    private static HashMap<String, Integer> totalPlayerPoints = new HashMap<>();
    // A map which has <Category, Total Count>
    private static HashMap<String, Integer> categoryPointMap = new HashMap<>();
    private static int totalPoints = 0;

    public static String lastMessage = "";
    
    public static void initPatterns() {
        currentPlayer = PartlySaneSkies.minecraft.getSession().getUsername();

        String str = PublicDataManager.getFile("constants/dungeons_player_rate_pattern_strings.json");
        if (str.equals("")) {
            return;
        }
        JsonObject patternJson = new JsonParser().parse(str).getAsJsonObject();
        JsonObject positivePatternsJson = patternJson.getAsJsonObject("positive_strings");

        for (Map.Entry<String, JsonElement> entry : positivePatternsJson.entrySet()) {
            positivePatterns.put(entry.getKey(), entry.getValue().getAsString());
        }


        
    }

    public static void rackPoints(String player, String category) {
        if (player.equalsIgnoreCase("You")) {
            player = currentPlayer;
        }

        for (String rank : PartlySaneSkies.RANK_NAMES) {
            player = player.replace(rank, "");
        }

        player = player.replaceAll("\\P{Print}", "");
        player = StringUtils.INSTANCE.stripLeading(player);
        player = StringUtils.INSTANCE.stripTrailing(player);

        // If the player has already been registered
        if (playerPointCategoryMap.containsKey(player)) {
            HashMap<String, Integer> individualPlayerPointMap = playerPointCategoryMap.get(player);
            // If the player already has this category registered
            if (individualPlayerPointMap.containsKey(category)) {
                individualPlayerPointMap.put(category, individualPlayerPointMap.get(category) + 1);
            }
            // If this category doesn't exist yet
            else {
                individualPlayerPointMap.put(category, 1);
            }

            totalPlayerPoints.put(player, totalPlayerPoints.get(player) + 1);
        }
        // If this player doesn't exist yet
        else {
            HashMap<String, Integer> individualPlayerPointMap = new HashMap<>();
            individualPlayerPointMap.put(category, 1);
            playerPointCategoryMap.put(player, individualPlayerPointMap);
            totalPlayerPoints.put(player, 1);
        }

        // If the categoryPointMap contains the category
        if (categoryPointMap.containsKey(category)) {
            categoryPointMap.put(category, categoryPointMap.get(category) + 1);
        }
        // If the categoryPointMap does not contain the category
        else {
            categoryPointMap.put(category, 1);
        }


        totalPoints ++;
    }
    public static String getDisplayString() {
        StringBuilder str = new StringBuilder();

        if (PartlySaneSkies.config.enhancedDungeonPlayerBreakdown == 0) { 
            for (Map.Entry<String, HashMap<String, Integer>> entry : playerPointCategoryMap.entrySet()) {
                String playerStr = "§d" + entry.getKey() + "  §9" + MathUtils.INSTANCE.round((double) totalPlayerPoints.get(entry.getKey()) / totalPoints * 100d, 0) +"%§7 | ";
                
                str.append(playerStr);
            }
            
            return (str.toString());
        }

        str.append("§a§nDungeon Overview:\n\n");
        for (Map.Entry<String, HashMap<String, Integer>> entry : playerPointCategoryMap.entrySet()) {
            String playerName = entry.getKey();
            StringBuilder playerStr = new StringBuilder("§d" + playerName + "§7 completed §d" + MathUtils.INSTANCE.round((double) totalPlayerPoints.get(playerName) / totalPoints * 100d, 0) + "%§7 of the dungeon.\n");
            if (PartlySaneSkies.config.enhancedDungeonPlayerBreakdown == 2) {
                playerStr.append("§2   Breakdown:\n");
                for (Map.Entry<String, Integer> entry2 : entry.getValue().entrySet()) {
                    playerStr.append("     §d").append(MathUtils.INSTANCE.round((double) entry2.getValue() / categoryPointMap.get(entry2.getKey()) * 100d, 0)).append("%§7 of ").append(entry2.getKey()).append("\n");
                }
            }
            
            str.append(playerStr);
        }

        str = new StringBuilder((str.toString()));



        return str.toString();
    }

    public static String getChatMessage() {
        StringBuilder str = new StringBuilder();

        str.append("Partly Sane Skies > ");
        
        for (Map.Entry<String, HashMap<String, Integer>> entry : playerPointCategoryMap.entrySet()) {
            String playerStr = entry.getKey() + "  " + MathUtils.INSTANCE.round((double) totalPlayerPoints.get(entry.getKey()) / totalPoints * 100d, 0) + "% | ";
            
            str.append(playerStr);
        }

        return str.toString();
    }


    public static void handleMessage(String message) {
        for (Map.Entry<String, String> entry : positivePatterns.entrySet()) {
            if (!StringUtils.INSTANCE.startsWithPattern(message, entry.getKey(), "{player}")) {
                continue;
            }

            rackPoints(StringUtils.INSTANCE.recognisePattern(message, entry.getKey(), "{player}"), entry.getValue());
        }
    }

    public static void reset() {
        categoryPointMap = new HashMap<>();
        totalPlayerPoints = new HashMap<>();
        playerPointCategoryMap = new HashMap<>();

        totalPoints = 0;
    }
    
    public static void reprintLastScore() {
        ChatUtils.INSTANCE.sendClientMessage(lastMessage, true);
    }

    public static void registerReprintCommand() {
        new PSSCommand("reprintscore", Arrays.asList("rps", "rs"), "Reprints the last score in a dungeon run", (s, a) -> {
            reprintLastScore();
        }).register();
    }

    // §r§fTeam Score: §r
    @SubscribeEvent
    public void onChatEvent(ClientChatReceivedEvent event) {
        if (!PartlySaneSkies.config.dungeonPlayerBreakdown) {
            return;
        }
        // If end of dungeon
        if (event.message.getFormattedText().contains("Catacombs Experience§r")) {
            final String string = getDisplayString();
            lastMessage = string;

            new Thread(() -> {
                try {
                    Thread.sleep(125);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
                PartlySaneSkies.minecraft.addScheduledTask(() -> {
                    if (string.equals("")) {
                        return;
                    }
                    ChatUtils.INSTANCE.sendClientMessage(string, true);
                });
            }).start();

            if (PartlySaneSkies.config.partyChatDungeonPlayerBreakdown) {
                PartlySaneSkies.minecraft.thePlayer.sendChatMessage("/pc " + getChatMessage());
            }
            reset();
            return;
        }

        if (event.message.getUnformattedText().contains("You are playing on profile:") || event.message.getFormattedText().contains("[NPC] §bMort§f: Here, I found this map when I first entered")) {
            reset();
            return;
        }
        
        handleMessage(event.message.getUnformattedText());
    }
}

File: src/main/kotlin/me/partlysanestudios/partlysaneskies/features/dungeons/PlayerRating.kt
//
// Written by Su386.
// See LICENSE for copyright and license notices.
//

package me.partlysanestudios.partlysaneskies.features.dungeons

import com.google.gson.JsonParser
import me.partlysanestudios.partlysaneskies.PartlySaneSkies.Companion.RANK_NAMES
import me.partlysanestudios.partlysaneskies.PartlySaneSkies.Companion.config
import me.partlysanestudios.partlysaneskies.PartlySaneSkies.Companion.minecraft
import me.partlysanestudios.partlysaneskies.commands.PSSCommand
import me.partlysanestudios.partlysaneskies.data.pssdata.PublicDataManager.getFile
import me.partlysanestudios.partlysaneskies.data.skyblockdata.IslandType
import me.partlysanestudios.partlysaneskies.events.SubscribePSSEvent
import me.partlysanestudios.partlysaneskies.events.data.LoadPublicDataEvent
import me.partlysanestudios.partlysaneskies.events.skyblock.dungeons.DungeonEndEvent
import me.partlysanestudios.partlysaneskies.events.skyblock.dungeons.DungeonStartEvent
import me.partlysanestudios.partlysaneskies.utils.ChatUtils.sendClientMessage
import me.partlysanestudios.partlysaneskies.utils.MathUtils.round
import me.partlysanestudios.partlysaneskies.utils.StringUtils.removeColorCodes
import me.partlysanestudios.partlysaneskies.utils.StringUtils.stripLeading
import me.partlysanestudios.partlysaneskies.utils.StringUtils.stripTrailing
import net.minecraftforge.client.event.ClientChatReceivedEvent
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent

object PlayerRating {
    private var currentPlayer = ""

    private var positiveRegexs = HashMap<Regex, String>()

    // A map that has the <Pattern, <Cause Category, Points>>
    private val playerPointCategoryMap = HashMap<String, HashMap<String, Int>>()

    // A map that has <Player, Total Points> Map
    private val totalPlayerPoints = HashMap<String, Int>()

    // A map which has <Category, Total Count>
    private val categoryPointMap = HashMap<String, Int>()
    private var totalPoints = 0

    private var lastMessage = ""

    @SubscribePSSEvent
    fun initPatterns(event: LoadPublicDataEvent) {
        currentPlayer = minecraft.session.username
        val str = getFile("constants/dungeon_player_rate_regex_strings.json")
        if (str == "") {
            return
        }
        val regexJson = JsonParser().parse(str).getAsJsonObject()
        val positiveRegexsJson = regexJson.getAsJsonObject("positive_strings")
        for ((key, value) in positiveRegexsJson.entrySet()) {
            positiveRegexs[key.toRegex()] = value.asString
        }
    }

    fun rackPoints(playerString: String, category: String) {
        var player = playerString
        if (player.equals("You", ignoreCase = true)) {
            player = currentPlayer
        }
        for (rank in RANK_NAMES) {
            player = player.replace(rank, "")
        }
        player = player.replace("\\P{Print}".toRegex(), "")
        player = stripLeading(player)
        player = stripTrailing(player)

        // If the player has already been registered
        if (playerPointCategoryMap.containsKey(player)) {
            val individualPlayerPointMap = playerPointCategoryMap[player]!!
            // If the player already has this category registered
            if (individualPlayerPointMap.containsKey(category)) {
                individualPlayerPointMap[category] = individualPlayerPointMap[category]!! + 1
            } else {
                individualPlayerPointMap[category] = 1
            }
            totalPlayerPoints[player] = totalPlayerPoints[player]!! + 1
        } else {
            val individualPlayerPointMap = java.util.HashMap<String, Int>()
            individualPlayerPointMap[category] = 1
            playerPointCategoryMap[player] =
                individualPlayerPointMap
            totalPlayerPoints[player] = 1
        }

        // If the categoryPointMap contains the category
        if (categoryPointMap.containsKey(category)) {
            categoryPointMap[category] = categoryPointMap[category]!! + 1
        } else {
            categoryPointMap[category] = 1
        }
        totalPoints++
    }

    private fun getDisplayString(): String {
        var str = StringBuilder()
        if (config.enhancedDungeonPlayerBreakdown == 0) {
            for ((key) in playerPointCategoryMap) {
                val totalPoints = ((totalPlayerPoints[key]?.toDouble() ?: 0.0) / totalPoints * 100.0).round(0)
                val playerStr = "§d$key  §9$totalPoints%§7 | "
                str.append(playerStr)
            }
            return str.toString()
        }
        str.append("§a§nDungeon Overview:\n\n")
        for ((playerName, value) in playerPointCategoryMap) {
            val totalPoints = ((totalPlayerPoints[playerName]?.toDouble() ?: 0.0) / totalPoints * 100.0).round(0)
            val playerStr = StringBuilder("§d$playerName§7 completed §d$totalPoints%§7 of the dungeon.\n")
            if (config.enhancedDungeonPlayerBreakdown == 2) {
                playerStr.append("§2   Breakdown:\n")
                for ((key, value1) in value) {
                    playerStr
                        .append("     §d")
                        .append((value1.toDouble() / (categoryPointMap[key] ?: 0) * 100.0).round(0))
                        .append("%§7 of ")
                        .append(key)
                        .append("\n")
                }
            }
            str.append(playerStr)
        }
        str = StringBuilder(str.toString())
        return str.toString()
    }

    private fun getChatMessage(): String {
        val str = java.lang.StringBuilder()
        str.append("Partly Sane Skies > ")
        for ((key) in playerPointCategoryMap) {
            val totalPoints = ((totalPlayerPoints[key]?.toDouble() ?: 0.0) / totalPoints * 100.0).round(0)
            val playerStr = "$key  $totalPoints% | "
            str.append(playerStr)
        }
        return str.toString()
    }

    private fun getSlackingMembers(): ArrayList<String> {
        val strList = ArrayList<String>()
        for ((key) in playerPointCategoryMap) {
            if (totalPlayerPoints[key]!! / (totalPoints * 1.0) > config.dungeonSnitcherPercent / 100f) {
                continue
            }
            strList.add("PSS Slacker Snitcher -> $key looks to be slacking. (This could be a mistake)")
        }
        return strList
    }

    private fun handleMessage(message: String) {
        for ((key, value) in positiveRegexs) {
            if (!(key.containsMatchIn(message.removeColorCodes()))) {
                continue
            }

            val (playerName) = key.find(message)?.destructured ?: return
            rackPoints(playerName, value)
        }
    }

    private fun reset() {
        categoryPointMap.clear()
        totalPlayerPoints.clear()
        playerPointCategoryMap.clear()
        totalPoints = 0
    }

    private fun reprintLastScore() {
        sendClientMessage(lastMessage, true)
    }

    fun registerReprintCommand() {
        PSSCommand(
            "reprintscore",
            mutableListOf("rps", "rs"),
            "Reprints the last score in a dungeon run",
        ) { reprintLastScore() }.register()
    }

    @SubscribePSSEvent
    fun onDungeonStart(event: DungeonStartEvent?) {
        if (!(config.dungeonPlayerBreakdown || config.dungeonSnitcher)) {
            return
        }
        reset()
    }

    @SubscribePSSEvent
    fun onDungeonEnd(event: DungeonEndEvent?) {
        if (!(config.dungeonPlayerBreakdown || config.dungeonSnitcher)) {
            return
        }
        val string = getDisplayString()
        lastMessage = string
        val chatMessageString = getChatMessage()
        val slackingMembers = getSlackingMembers()

        Thread {
            try {
                Thread.sleep((config.dungeonPlayerBreakdownDelay * 1000).toLong())
            } catch (e: InterruptedException) {
                e.printStackTrace()
            }
            minecraft.addScheduledTask {
                if (string == "") {
                    return@addScheduledTask
                }
                sendClientMessage(string, true)
                if (config.partyChatDungeonPlayerBreakdown) {
                    minecraft.thePlayer.sendChatMessage("/pc $chatMessageString")
                }
            }
            if (config.dungeonSnitcher) {
                for (str in slackingMembers) {
                    try {
                        Thread.sleep(750)
                    } catch (e: InterruptedException) {
                        throw RuntimeException(e)
                    }
                    minecraft.addScheduledTask {
                        minecraft.thePlayer.sendChatMessage("/pc $str")
                    }
                }
            }
        }.start()
        reset()
    }

    @SubscribeEvent
    fun onChatEvent(event: ClientChatReceivedEvent) {
        if (!(config.dungeonPlayerBreakdown || config.dungeonSnitcher)) {
            return
        }
        if (event.message.unformattedText.contains("You are playing on profile:")) {
            reset()
            return
        }
        if (!IslandType.CATACOMBS.onIsland()) {
            return
        }

        handleMessage(event.message.unformattedText)
    }
}
